packages/shared/src/models-hdf5.ts

```diff
@@ -164,7 +164,7 @@ export type ArrayValue<T extends DType> =
 export type Value<D extends Dataset> = D['shape'] extends ScalarShape
   ? Primitive<D['type']>
   : D['shape'] extends ArrayShape
-  ? Primitive<D['type']>[]
+  ? ArrayValue<D['type']>
```

Comment:
Another very important change, that I had mentioned in my previous PR, which lets providers return typed arrays (from TS' point of view).